include/exec/system_context.hpp

```diff
@@ -0,0 +1,433 @@
+      __exec_system_scheduler_interface* __scheduler_{nullptr};
+    };
+
+    /// Connects `__self` to `__r`, returning the operation state containing the work to be done.
+    template <class __R>
+    friend auto tag_invoke(stdexec::connect_t, system_sender&& __self, __R&& __r) noexcept(
+      std::is_nothrow_constructible_v<std::remove_cvref_t<__R>, __R>)
+      -> __op<system_sender, std::remove_cvref_t<__R>> {
+
+      return {std::move(__r), __self.__scheduler_};
+    }
+
+    /// Gets the environment of this sender.
+    friend __system_scheduler_env
+      tag_invoke(stdexec::get_env_t, const system_sender& __self) noexcept {
+      return {__self.__scheduler_};
+    }
+
+    /// The underlying implementation of the system scheduler.
+    __exec_system_scheduler_interface* __scheduler_{nullptr};
+  };
+
+  /// The state needed to execute the bulk sender created from system context.
+  template <stdexec::sender __Previous, std::integral __Size, class __Fn, class __R>
+  struct __bulk_state {
+    /// The sender object that describes the work to be done.
+    system_bulk_sender<__Previous, __Size, __Fn> __snd_;
+    /// The receiver object that receives completion from the work described by the sender.
+    __R __recv_;
+    /// The arguments passed from the previous receiver to the function object of the bulk sender (type-erased).
+    char* __arguments_data_{nullptr};
+  };
+
+  /// Receiver that is used in "bulk" to connect toe the input sender of the bulk operation.
+  template <stdexec::sender __Previous, std::integral __Size, class __Fn, class __R>
+  struct __bulk_intermediate_receiver {
+    /// Declare that this is a `receiver`.
+    using receiver_concept = stdexec::receiver_t;
+
+    /// The type of the object that holds relevant data for the entire bulk operation.
+    using __bulk_state_t = __bulk_state<__Previous, __Size, __Fn, __R>;
+
+    /// Object that holds the relevant data for the entire bulk operation.
+    __bulk_state_t& __state_;
+
+    /// Invoked when the previous sender completes with a value to trigger multiple operations on the system scheduler.
+    template <class... __As>
+    friend void tag_invoke(
+      stdexec::set_value_t,
+      __bulk_intermediate_receiver&& __self,
+      __As&&... __as) noexcept {
+      // Heap allocate input data in shared state as needed
+      auto __size = sizeof(std::tuple<__As...>);
+      __self.__state_.__arguments_data_ = new char[__size];
+      new (__self.__state_.__arguments_data_) std::tuple<__As...>{__as...};
+      // TODO: figure out from the completion signal the size of the input data, and preallocate this in the shared state.
+
+      auto __type_erased_fn = [](void* __state_, long __idx) {
+        auto* __state = static_cast<__bulk_state_t*>(__state_);
+        std::apply(
+          [&](auto&&... __args) { __state->__snd_.__fun_(__idx, __args...); },
+          *reinterpret_cast<std::tuple<__As...>*>(__state->__arguments_data_));
+      };
```

Comment:
so there are going to be `shape` indirect function calls? i'm curious if there's a better way.

Reply:
I doubt there is a better way. In the common case, each call will be made on a different thread.

In general, to schedule work, we need 1 + N calls across the external interface (possible to a different library), where N is the number of threads we need.